### PR TITLE
Fix reading of empty BinTableHDU when stored in a gzip-compressed file

### LINKED ISSUE #8802: I/O for empty BinTableHDU
A bit of a low-priority corner case:

When an empty `BinTableHDU` is written to a file and then the file is reopened, the result depends on whether or not the written file was gzipped:

```
import numpy
from astropy.io import fits

ofile = 'emptytable.fits'
fits.HDUList([fits.PrimaryHDU(),
              fits.ImageHDU(data=numpy.zeros(100).reshape(10,10)),
              fits.BinTableHDU(header=None, data=None, name='TEST')]).writeto(ofile, overwrite=True)

hdu = fits.open(ofile)
print(hdu['TEST'].data is None)
print(hdu['TEST'].data.size == 0)

ofile = 'emptytable.fits.gz'
fits.HDUList([fits.PrimaryHDU(),
              fits.ImageHDU(data=numpy.zeros(100).reshape(10,10)),
              fits.BinTableHDU(header=None, data=None, name='TEST')]).writeto(ofile, overwrite=True)

hdu = fits.open(ofile)
print(hdu['TEST'].data is None)
```

The first two print statements work fine, but the last one faults with the following:
```
Traceback (most recent call last):
  File "binio.py", line 19, in <module>
    print(hdu['TEST'].data is None)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/astropy/utils/decorators.py", line 744, in __get__
    val = self.fget(obj)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/astropy/io/fits/hdu/table.py", line 406, in data
    data = self._get_tbdata()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/astropy/io/fits/hdu/table.py", line 172, in _get_tbdata
    self._data_offset)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/astropy/io/fits/hdu/base.py", line 475, in _get_raw_data
    return self._file.readarray(offset=offset, dtype=code, shape=shape)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/astropy/io/fits/file.py", line 338, in readarray
    data = _array_from_file(self._file, dtype, count)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/astropy/io/fits/util.py", line 585, in _array_from_file
    array = np.frombuffer(s, dtype=dtype, count=count)
ValueError: itemsize cannot be zero in type
```

It's not hard to work around this (just don't write extensions with empty binary tables), but I thought I would bring it up.

Python 3.6.2
numpy==1.16.2
astropy==3.1.2

Let me know if you need any other versions.

Thanks!

### PR DESCRIPTION
**Pull request by saimn**
Fix #8802.

### CODE REVIEW

File: astropy/io/fits/util.py [hunk 1]
```diff
@@ -583,10 +583,13 @@ def _array_from_file(infile, dtype, count):
         # their underlying file object, instead of the decompressed bytes
         read_size = np.dtype(dtype).itemsize * count
         s = infile.read(read_size)
-        array = np.frombuffer(s, dtype=dtype, count=count)
-        # copy is needed because np.frombuffer returns a read-only view of the
-        # underlying buffer
-        array = array.copy()
+        if s == b'':
```
Comment (pllim) [inline astropy/io/fits/util.py]: I am still scarred from the Python 2 to 3 transition. Are you sure this works fine in Python 2 as well, given you milestoned it to 2.x?

File: astropy/io/fits/util.py [hunk 2]
```diff
@@ -583,10 +583,13 @@ def _array_from_file(infile, dtype, count):
...
+            array = np.ndarray(buffer=b'', dtype=np.dtype([]), shape=(0,))
```
Comment (mhvk) [inline astropy/io/fits/util.py]: I would have hoped that you would not have to do an `if` statement at all. Did you try:
```
array = np.ndarray(buffer=s, dtype=dtype, shape=(count,))
```

Comment (saimn) [reply astropy/io/fits/util.py]: I was not sure while reading your comment ;), I will try!

File: astropy/io/fits/util.py (hunk 1 above)
Comment (saimn) [reply astropy/io/fits/util.py]: I think it should work, but hopefully we can get rid of the `if` with Marten's suggestion.

**Comment by bsipocz** [comment]
Rebased this  and I go ahead as merged given that the review comments above have been addressed, and tests passed before rebase (conflict in changelog only).

**Comment by bsipocz** [comment]
Thanks @saimn!

**Comment by bsipocz** [comment]
This one is  just not working with LTS: https://travis-ci.org/astropy/astropy/builds/546000389, so I remilestone to v3.2.1.

**Comment by mhvk** [comment]
Seems fine to re-milestone given that this addressed a corner case.

**Pull request merged**